### DISCUSSION TITLE
CJS support

### DISCUSSION THREAD

**Discussion by notaphplover** [faker-js/faker#3103] [General]
While discussing https://github.com/faker-js/faker/issues/3087 it came to my attention [CJS might stop being supported on the next major release](https://github.com/faker-js/faker/issues/3087#issuecomment-2332009238).

Is that so? Are you open to discuss this topic?

My cup of tea would be continue supporting CJS, and since other users (even faker contributors) share the same concerns, I though it might be worth discussing this topic.

**Comment by xDivisionByZerox** [comment]
First things first, thank you for opening a separate discussion to keep the context clear and minimal.

---

Just to be transparent about this topic:
While not specifically mentioned in any issue/PR or documentation, it is part of the description for [the milestone for v10](https://github.com/faker-js/faker/milestone/16).

**Reply by notaphplover** [reply]
Thank you so much! All right. May we know the rationale behind this decision?

While I absolutely see benefits in promoting the use of ESM, I actually find pretty much impossible to go for it in some common use cases. Today it's not possible to rely on ESM without using a bundler so you can use libraries too good to not to be used, like `jest`. Even with the introduction of a bundler, this process is long, painful and, sometimes, it's not worth it. When I picture things these way, migrating to faker@10 might not be a great idea for part of the userbase. I would prefer to avoid this, keeping CJS compatibility, but I understand you wouldn't drop CJS support without a reason to do so. That's the reason to ask, so we can understand your concerns and, maybe, come to a better solution for eveyone.

**Reply by Shinigami92** [reply]
Just to bring also my perspective: v10 will earliest come when Node v18 is non LTS anymore (around Q2 2025)
FakerJS has around 5mb of translation files, and right now these are doubled, one for cjs and one for esm.
We want not only to get rid of these cjs translations, but also benefit from new stuff like `import.meta.dirname` (Node v20.11+)

**Reply by xDivisionByZerox** [reply]
Please note that `import.meta.dirname` is a bad example for dropping CJS for users as this would be used in development only. Additionally, this is related to the version of Node and while only available in ESM, it doesn't provide any argument for CJS support for users.

---

Regarding the translations, I once again advocate for dedicated npm packages under the faker namespace. For example:

- `@faker-js/locale-en`
- `@faker-js/locale-en-US`
- `@faker-js/locale-de`
- ...

While this still requires dedicated builds for CJS and ESM, it would significantly reduce the amount of data to download for each user as only the ESM and CJS files for the dedicated locale would be downloaded.

I want to emphasize that the above suggestion is an alternative route meant for constructive feedback - not as a preference from myself.  
In a perfect world, this discussion wouldn't even happen, to be honest.

**Reply by Shinigami92** [reply]
I would like more to sometimes backport features to v9 as continuing this madness in v10 🫠 
Just kill cjs once and for all, I want to benefit from future things and not legacy software

**Reply by prisis** [reply]
I see the only problem with cjs is that a lot of bundler dont do esm and cjs exports correctly, and the size is doubled.

In the end you writing typescript and for the rest the bundler should do the job.

I like the idea to split faker into translation packages, where are all translations needed in a project?

**Reply by notaphplover** [reply]
So, our main concerns here seem to be:

1. The additional complexity as a result for supporting both CJS and ESM modules.
2. Package sizes.

Is that so?

Regarding the additional complexity of supporting both ESM and CJS modules, it's a trade of. I agree removing CJS support would led to a much simpler library. But CJS support is a feature demanded in your userbase. If backend projects rely on faker to generate fake data, removing CJS support is a problem. It has already happened in the past. Consider [got weekly downloads](https://www.npmjs.com/package/got?activeTab=versions): only around 15% of then target `got@12` or greater. Can you wonder why? Simple: `got@12` droped CJS support and it's supposed to be a nodeJS library. That also explains why `got@11` is, by far, the most downloaded version set. I think this is an example of what removing CJS support means in the backend ecosystem. In a perfect world, we all would use ESM, of course. But I'm afraid it's not the case.

Regarding package size, is that important?

faker is mostly used as dev dependency. Even as non dev dependency, if package size is a concern, doesn't any sort of tree shaking approach simply deal with this problem? I mean, If I use faker as a non dev dependency, I'd be puting my faith in a bundler good enough to know I'm targeting ESM so CJS modules from faker should not be transpiled nor bundled since faker is already providing ESM ones.

For the sake of a constructive feedback, we can simply ignore what I said and focus on the package size. Even in this case, I wonder if droping CJS support is the solution to your problem. Maybe having a big package size is a problem that demands a solution, but it doesn't necessarily mean that dropping CJS support is the right approach. I do like @xDivisionByZerox approach of providing locale packages (which maybe would be installed as peer dependecies?), but my 5 cents to this discussion would be: why are we coding data as code? Maybe we are considering an unnecessary duplication of code because, in the end, translations are data rather than code?

An npm package has modules, but an npm package doesn't necessarily include only modules. You can include any file as well. You can include json, as an example, and it's perfectly fine to have a translation package with json translations and import them / require them. It's an alterative approach that would deduplicate translations.

**Reply by xDivisionByZerox** [reply]
> [...] why are we coding data as code? Maybe we are considering an unnecessary duplication of code because, in the end, translations are data rather than code?
> 
> An npm package has modules, but an npm package doesn't necessarily include only modules. You can include any file as well. You can include json, as an example, and it's perfectly fine to have a translation package with json translations and import them / require them. It's an alterative approach that would deduplicate translations.

FYI there was #1287 that experimented with this approche.

**Reply by Shinigami92** [reply]
As @xDivisionByZerox posted, I for myself already suggested this years ago! And for me we now running in circles. I'm so tired and sick of cjs that I think I would leave Faker behind me if we don't drop cjs support in v10.
If some of the community still need cjs for legacy project reasons, they are totally safe to pin v8 or v9 and stick to it. Faker does not even have direct dependencies and so they won't introduce security risks as far as I can tell.
I have zero problems with leaving the cjs community behind and move on, because we have this situation for years now in the ecosystem and if big libraries with millions of downloads a week don't enforce and move on, this situation will stay for years.

**Reply by ST-DDT** [reply]
There are certain tradeofs to everything.

- if we switch to json data:
  - our bundle size might be smaller
  - there is some additional complexity related to type checking of the json data
  - the theoretical min bundle size of our users will be bigger (because data objects aren't tree-shaken). See also nano bound standalone module functions #2667 (this depends a bit on which files are jsonified)
- as for cjs:
  - I'm not against retaining support for it, but it must be worth it
  - dropping support for it would reduce our build process complexity by a lot, we could potentially switch back to plain tsc. We could then invest that complexity into new features such as nano bound module functions.
  - we are currently experiencing build issues related to cjs and moduleResolution, so it causes perceived pain. If you/the community could review and test the related PRs, that would restore some confidence in them. #3087

**Reply by notaphplover** [reply]
Thank you so much for all your feedback. It's really appreciated.

If json files cannot be tree shaked, I would avoid then in favor of module duplication.

I've had a look at #2667, it looks great, it was quite instructive tbh.

Regarding the use of tsc, may I know what's blocking you from using it? I wasn't expecting cjs support to be part of the problem, I was expecting whatever tsc based config which works on ESM can simply be replicated on CJS unless some few exceptions.
